### ISSUE TITLE
Execute code after all retries have been exhausted

### ISSUE THREAD

**Issue by SeanFarrow** [App-vNext/Polly#123]
Hi,

I have the following policy:
Policy.Handle<MissingFeedException>().Or<InvalidFeedXmlException>().WaitAndRetryAsync(3, (attempt) =>TimeSpan.FromSeconds(0.1 *Math.Pow(2, attempt)));
Is there a way to execute code after all retries have been exhausted?
Cheers
Sean.

**Comment by reisenberger** [comment]
@SeanFarrow .  At the moment there isn't a way within Polly to execute specific code after all retries have been exhausted. However, this is on the Roadmap as a feature we plan to add.  It's described within #80 and the [Roadmap](https://github.com/App-vNext/Polly/wiki/Polly-Roadmap) as a **Fallback** policy.

+ The Fallback policy would specify a lambda to execute (eg to provide an alternative value; or take other remedial action) if the main policy (retry/ circuit-breaker/ whatever) still eventually threw.
+ The Fallback policy could be wrapped around any other policy, as described in the **Pipeline** in the [Roadmap](https://github.com/App-vNext/Polly/wiki/Polly-Roadmap), so that Fallback could eventually be combined with a wider range of transient-fault-handling techniques: cache, retry/circuit-breaker combined, throttle, timeout etc.

Please feel free to have a read of the proposed feature and let us know if you have any feedback!

**For now**, to execute code after all retries have been exhausted, you can either try-catch around your `.ExecuteAsync(...)` call to catch the final exception.  Or, use `.ExecuteandCapture(...)` and identify if there was a final exception from the returned `PolicyResult`.

Hope this helps! (and let us know if there is anything else you need assistance with)

**Comment by SeanFarrow** [comment]
Hi,

Just read the roadmap, the piping feature, cache feature and fall back policy are definitely of interest.
I’m happy and have some time to help develop these. I’m currently writing a system that must be completely resilient so poly will form the core of this.
Cheers
Sean.

**Comment by reisenberger** [comment]
@SeanFarrow  Great.  There is a lot of material on the Roadmap: contributions will always be welcome.  

The App-vNext team has some relatively developed ideas about how some of this functionality can be implemented (for example from similar in-house experience without Polly), but, like any OSS project, ideas of course also up for debate as the work evolves.

@SeanFarrow  Would be very happy to hear more about the project you are looking to slot this into (to whatever extent you able to share). Helps us better understand perspective you bring (all new perspectives valuable...) and angle on any issues.

Are there particular timescales you are looking to meet for the project you are working on?


At the moment next step is to unblock #14; relatedly, we probably need to split the Polly codebase between non-generic `Policy` and a generic `Policy<TResult>`, for most of the new-policy features on the wider resilience roadmap.  Many of these (Fallback; Cache; Handle Return Values) depend on operating to return a common `TResult`, and will compose as a Pipeline more coherently from a strongly-typed generic `Policy<TResult>`.  I am now well embarked on this (after a regular range of async refinements and circuit-breaker extensions since App-vNext took over the project Dec15), but it's also a relatively large piece of work.

@SeanFarrow  I have a couple of major annual meetings to prepare for this week Wednesday/Thursday/Fri; hope to set out more about vision for these features after that.

**Comment by SeanFarrow** [comment]
Hi,

I can share what I’m doing provate with an NDA if that helps.
Please contact me privately to discuss.
Cheers
Sean.

**Comment by reisenberger** [comment]
Closing as duplicate of #80
